text/0006-equivocation-catastrophes.md

```diff
@@ -0,0 +1,111 @@
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+The procedure for handling the equivocation catastrophes will have to consist of two main parts: what the software is expected to do in such a case, and what are the manual actions required to proceed.
+
+The obvious first thing that should happen in the case of an equivocation catastrophe is that the software should alert the users about the situation. The node software is aware of the configured FTT, so when it detects equivocations, it can notice if the total weight of the equivocators exceeds the current FTT. When it is exceeded, the users should be made aware of it.
+
+In order to make sure that as many validators as possible become aware of all the equivocations eventually, the operation of the nodes shouldn't stop. Finalizing blocks should be halted, as finalization ceases to mean anything when it can be reverted, but the nodes should continue gossiping votes.
+
+The nodes could also output the hash of the last block finalized before FTT has been exceeded and all the transactions included in the later blocks, and the list of equivocators. These will be useful for the manual steps.
+
+The manual actions afterwards will consist of two main steps:
+
+1. Reaching social consensus about the new starting point,
+2. Restarting the network with the new starting point.
+
+We will go into more detail about these steps below.
+
+### Social consensus
+
+For the initial version of this CEP, we assume that we have some kind of platform that enables open communication between validators and CasperLabs. The chosen platform should allow for storing, linking, referencing and displaying the whole public process and communication. We also should make sure that the content is in a format that can be displayed elsewhere.
+
+The first essential step will be determining the starting point for the network after the restart. The suggested approach here would be to agree upon the new initial global state, new set of validators and restart the network as if it was starting from scratch.
+
+A desirable goal for the initial global state would be to revert as few transactions from all the forks as possible, in order to limit the damage. Every transaction that was included in a block which appeared to be finalized for a time could have caused real world consequences (like, for example, a merchant sending ordered goods). Reverting any such transaction might potentially cause a loss to a user.
+
+In order to minimize losses, the following approach is proposed: first, choose the last finalized block as the starting point. If the equivocators caused multiple forks to appear, choose the tip of one of the forks. The exact criteria of choice will be left to the community to decide.
```

Comment:
> the last finalized block

That is not objective anymore, after the equivocation catastrophe, if only the output of our standard finality detector is taken into account.
What _would_ be objective, after syncing the full DAG, would be the last block still appearing finalized with a 1% FTT, but we'd need to add some code to find that block.

Reply:
Hmm... Would that have to be 1% FTT, or _any_ common FTT would be sufficient?
If it's not any FTT, then I guess the reason would be that FTTs exceeded by the weight of equivocating validators would not do. In such a case, wouldn't we have to choose the last block appearing finalized with some chosen FTT higher than the weight of equivocators?

Reply:
The way we currently calculate a block's FTT, equivocators are already excluded from any summit.
So once the catastrophe has happened and we've fully synced up our protocol states — let's call that the "final state" —, we'll all see the _f_ equivocators as faulty.
* Any block seen finalized with an FTT _> f_ has the property that nobody ever saw a conflicting block finalized (even before syncing up). The last such block might still be older than some blocks that _some_ observers saw finalized with FTT _> f_ (or above the configured threshold) before the catastrophe.
* There is a unique last block seen as finalized with FTT _> 0_ in that final state. That block is a descendant of all blocks that were ever observed finalized with FTT _> f_ by anyone.

If that's correct (it is, isn't it? :thinking:), I guess both would be canonical choices: The last block still seen with FTT _> f_, and the last block seen with FTT _> 0_ in the final state.
I think maybe the latter is preferable, because it means at least you know that whatever you observed with FTT _> f_ is still considered final. It gives you the guarantee that, if you locally require FTT 80% to consider something final, and fewer than 80% equivocate, even if the network considers that a catastrophe, your "finalized" blocks are still finalized after the restart.

Reply:
> The last such block might still be older than some blocks that some observers saw finalized with FTT > f (or above the configured threshold) before the catastrophe.

I don't get this part. If FTT > f and the protocol states are fully synced, then all validators see the same blocks as finalized, and no finalization could be reverted, right? So why would the last finalized block in such a situation be older than some blocks some observers saw as finalized before the catastrophe?
Or is it a typo and you meant FTT < f in that last part?

Reply:
What I mean is that equivocations can now _reduce_ the FTT for already finalized blocks. If a block is finalized with 35%, and then 5% equivocate, it can happen that afterwards, that block's FTT is only 30%.

So it could happen that you already observed a block with 80% FTT, then 70% equivocated, so now that block looks like it has only 10%, even though some participants observed it as finalized with 80%.

I think it would be good if such a block would then be guaranteed to remain finalized.

Reply:
Wait, so how does that work now? We still have some concrete FTT defined in the chainspec. Say it's 40%. We see a block finalized with this FTT, then 30% equivocate. Does that mean that the block is now no longer finalized with FTT 40%? Even though it's FTT hasn't been exceeded?

Reply:
Kind of. But we don't detect that and we continue considering it finalized anyway.
The idea is that it will almost certainly reach the configured FTT again, unless the equivocators actually exceed that limit. So instead of us doing something about that block not having sufficient FTT anymore, we just wait, and assume that eventually, everyone will see it finalized again.

Reply:
Hmm... That sounds kind of unsafe. Could it be possible for malicious actors to propose blocks and vote in such a way that we'd have two competing blocks, one is finalized "temporarily", then they equivocate just enough for it not to be finalized anymore with the original FTT and eventually the other block is finalized with that FTT instead?

Reply:
No, we can't (I think :sweat_smile:).
The only way to finalize (with FTT _> 0_) a block conflicting with a block that anyone ever saw as finalized with FTT _f_ is if _f_ validators equivocate.